src/blocks/Pairing.sol

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.16;
+
+import "src/blocks/grumpkin/Bn256.sol";
+
+/**
+ * @title Pairing Library
+ * @notice Provides functionalities for elliptic curve pairing operations, specifically for BN256 curve. Based on:
+ *         https://gist.github.com/chriseth/f9be9d9391efc5beb9704255a8e2989d
+ * @dev This library is essential for cryptographic operations that require pairing checks.
+ */
+library Pairing {
+    // Represents a point on G1 (first group of BN256 curve).
+    struct G1Point {
+        Bn256.Bn256AffinePoint inner;
+    }
+    // Represents a point on G2 (second group of BN256 curve).
+
+    struct G2Point {
+        uint256[2] X;
+        uint256[2] Y;
+    }
+
+    /**
+     * @notice Computes the pairing check of two sets of points.
+     * @param p1 Array of G1 points.
+     * @param p2 Array of G2 points.
+     * @return True if the pairing check passes, false otherwise.
+     * @dev For example pairing([P1(), P1().negate()], [P2(), P2()]) should
+     *      return true.
+     */
+    function pairing(G1Point[] memory p1, G2Point[] memory p2) internal returns (bool) {
+        require(p1.length == p2.length);
+        uint256 elements = p1.length;
+        uint256 inputSize = elements * 6;
+        uint256[] memory input = new uint256[](inputSize);
+        for (uint256 i = 0; i < elements; i++) {
+            input[i * 6 + 0] = p1[i].inner.x;
+            input[i * 6 + 1] = p1[i].inner.y;
+            // To be compatible with Rust, it is necessary to reverse the field element encoding
+            input[i * 6 + 2] = p2[i].X[1];
+            input[i * 6 + 3] = p2[i].X[0];
+            input[i * 6 + 4] = p2[i].Y[1];
+            input[i * 6 + 5] = p2[i].Y[0];
+        }
+        uint256[1] memory out;
+        assembly {
+            let success := call(sub(gas(), 2000), 8, 0, add(input, 0x20), mul(inputSize, 0x20), out, 0x20)
+            switch success
+            case 0 { revert(0, 0) }
```

Comment:
The original mentions returning invalid here to make gas estimation work. Does it perform well in the case of this PR?

Reply:
TBH, I don't completely understand this pairing precompile invocation code. With `invalid()` it still produces `OutOfGas` error in my current settings. Will need to spend some time to figure it out. Anyway, it is indeed true that we need to have some correct gas cost value on test failure